platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py

```diff
@@ -0,0 +1,491 @@
+        }    
+            
+        self.debug_print("Calling hash calculator for ECMP")    
+        hash_result = self.call_hash_calculator({'ecmp_hash': ecmp_hash})
+        ecmp_hash_result = hash_result['ecmp_hash']
+        index = ecmp_hash_result['ecmp_index']  
+        
+        self.debug_print("Next hop index to which trafic will egress: {}".format(index))
+        return index  
+
+    # Get index in LAG memebrs array from which packet will egress
+    def get_lag_member_index(self, lag_size, flood_case = False):
+        logical = self.get_ingress_port_logical_idx()
+
+        lag_hash = {
+            "ingress_port": str(hex(logical)),
+            "packet_info": self.packet['packet_info'],
+            "lag_size": lag_size,
+        }
+    
+        self.debug_print("Calling hash calculator for LAG, flood case {}".format(True if flood_case else False))
+        hash_result = self.call_hash_calculator({"lag_hash": lag_hash})
+        lag_hash_result = hash_result["lag_hash"]
+        
+        if flood_case:
+            index = lag_hash_result['lag_mc_index']
+        else:
+            index = lag_hash_result['lag_index']
+
+        self.debug_print("Lag member index from which trafic will egress: {}".format(index))
+        return index
+        
+    # Get LAG memebr from which packet will egress
+    def get_lag_member(self, logical, flood_case = False):
+        lag_members = sx_lag_port_group_get(self.handle, logical)
+        lag_members.sort()
+        
+        member_index = self.get_lag_member_index(len(lag_members), flood_case)
+        lag_member = lag_members[member_index]
+        
+        self.debug_print("Lag member from which trafic will egress: {}".format(lag_member))
+        return lag_member
+        
+    def call_hash_calculator(self, input_dict):
+        with open(HASH_CALC_INPUT_FILE, "w") as outfile:
+            json.dump(input_dict, outfile)
+        
+        out = exec_cmd([HASH_CALC_PATH, '-c', HASH_CALC_INPUT_FILE, '-o', HASH_CALC_OUTPUT_FILE, '-d'])
+        self.debug_print ("Hash calculator output:\n{}".format(out))
+    
+        with open(HASH_CALC_OUTPUT_FILE, 'r') as openfile:
+            output_dict = json.loads(openfile.read())
+            
+        return output_dict
+                    
+    def get_active_vrids(self):
+        self.vrid_list = sx_get_active_vrids(self.handle)
+        
+    def validate_ipv4_header(self, header):
+        for ip in ['sip', 'dip']:
+            if ip in header and is_ip_valid(header[ip], IP_VERSION_IPV4) == False:
+                raise ValueError("Json validation failed: invalid IP {}".format(header[ip]))
+
+    def validate_ipv6_header(self, header):
+        for ip in ['sip', 'dip']:
+            if ip in header and is_ip_valid(header[ip], IP_VERSION_IPV6) == False:
+                raise ValueError("Json validation failed: invalid IP {}".format(header[ip]))
+            
+    def validate_layer2_header(self, header):
```

Comment:
Should ethertype, vlan and other L2 fields be validated as well?

Reply:
By design, Ethertype, vlan and other L2 fields are being validated for their types only.
Each field is defined in PACKET_SCHEME as a number or as a string and validation is being performed using python3-jsonschema module.